server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4662,35 +4660,31 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
-                                    Map<String, String> userData = new HashMap<>(infosSnapshot.getUserData());
-                                    userData.put(TRANSLOG_UUID_KEY, localSegmentInfos.userData.get(TRANSLOG_UUID_KEY));
-                                    infosSnapshot.setUserData(userData, false);
-                                }
-                                storeDirectory.deleteFile(localMaxSegmentInfos.get());
+                    if (this.shardRouting.primary()) {
+                        long processedLocalCheckpoint = Long.parseLong(infosSnapshot.getUserData().get(LOCAL_CHECKPOINT_KEY));
+                        // Following code block makes sure to use SegmentInfosSnapshot in the remote store if generation differs
+                        // with local filesystem. If local filesystem already has segments_N+2 and infosSnapshot has generation N,
+                        // after commit, there would be 2 files that would be created segments_N+1 and segments_N+2. With the
+                        // policy of preserving only the latest commit, we will delete segments_N+1 which in fact is the part of the
+                        // latest commit.
+                        Optional<String> localMaxSegmentInfos = localSegmentFiles.stream()
+                            .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
+                            .max(Comparator.comparingLong(SegmentInfos::generationFromSegmentsFileName));
+                        if (localMaxSegmentInfos.isPresent()
+                            && infosSnapshot.getGeneration() < SegmentInfos.generationFromSegmentsFileName(localMaxSegmentInfos.get())
+                                - 1) {
+                            // If remote translog is not enabled, local translog will be created with different UUID.
+                            // This fails in Store.trimUnsafeCommits() as translog UUID of checkpoint and SegmentInfos needs
+                            // to be same. Following code block make sure to have the same UUID.
+                            if (indexSettings.isRemoteTranslogStoreEnabled() == false) {
+                                SegmentInfos localSegmentInfos = store.readLastCommittedSegmentsInfo();
+                                Map<String, String> userData = new HashMap<>(infosSnapshot.getUserData());
+                                userData.put(TRANSLOG_UUID_KEY, localSegmentInfos.userData.get(TRANSLOG_UUID_KEY));
+                                infosSnapshot.setUserData(userData, false);
                             }
-                            store.commitSegmentInfos(infosSnapshot, processedLocalCheckpoint, processedLocalCheckpoint);
+                            storeDirectory.deleteFile(localMaxSegmentInfos.get());
                         }
-                    } else {
-                        finalizeReplication(infosSnapshot);
```

Comment:
As we are removing this, how do we update the segments for replica?

Reply:
I believe this is no longer required because we are no longer invoking this method on a started engine.  Only during store recovery / peer recovery before the engine is opened which will read from disk.  Is this correct @dreamer-89 ?

Reply:
Till now, we have only one method to download the segments from remote store. There is nothing stopping us to have download logic at two different places. Just want to make sure that this does not break any assumptions.

Reply:
is there any benefit of forking the logic here? just thinking out loud, would it make sense to reuse the `IndexShard.copySegmentFiles` for our case which already take a list of files to download from remote store?

Reply:
@sachinpkale : Thanks for raising this. This is a valid concern. There are couple of reasons for using separate download logic for regular rounds of segment replication
1. [Existing logic](https://github.com/opensearch-project/OpenSearch/blob/4319f2b571c82d7ec5608ffdc31ca4105961d064/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L4766) iterates through all files on remote and copies to local (on replica) which are missing. The missing files are identified by listing all directory files. 
  a.  **Non-performant**. For round of segment replication, the missing files is already known (evaluated before requesting files). Thus not needed to iterate through all files.
  b.  **Code duplicacy**. Missing files is already identified inside SegmentReplicationTarget and thus would be duplicate to adopt existing method.
2. The method is not generic enough to be used by different consumers. e.g. download specific files, don't perform check for file exists etc

CC @mch2

Reply:
The issue here is that this method is not simply downloading segments and has side effects. I think it makes sense to have the actual download logic in one place in the store from a list of needed files that both segrep and recovery can reuse. The segrep flow already has opinions on how to handle diffs etc that are different from recovery.